### PR TITLE
Add clock functions

### PR DESCRIPTION
**Pull request by guansong**

### CODE REVIEW

**Comment by b-sumner** [comment]
Most of these functions can and should be implemented in C using the irif header.
I doubt any user code directly calls those nvvm.barrier instrinsics but if it does we need to find out why and fix it.

**Comment by guansong** [comment]
I need to bring Greg's attention here and needs his opinion on this.

I put two changes in this PR, one in CL one in LL. They both come from our nvptx intrinsic wrappers.  Ideally the barrier should be written in CL as well (I split the original wrapper file just to maximize the usage of CL). But the naming convention there (with a .) prevents me to do so.

Brian suggested to fix the original place of using them. I see we have lot of similar functions used that way. Do we know those are from users or compiler generated code?

**Comment by b-sumner** [comment]
I suppose it is possible the front end is directly generating those intrinsic calls.  In that case you can still implement in C by first declaring the function and then implementing it, like this

extern void __llvm_nvvm_barrier0(void) __asm("llvm.nvvm.barrier0");
void __llvm_nvvm_barrier0(void) { /* implementation here */  }

I am not really understanding why those barrier functions you listed do not have an additional "llvm." prefix. And I have no idea what these functions are actually supposed to do.  Your patch shows barrier1 has two i32 parameters.  What do they do?  Can you find the locations in the original source code corresponding to those calls?

**Comment by guansong** [comment]
I see. 

How about the clock routines? Are those OK?

I will ask Greg more for the frontend part. My guess is that it maybe either Greg or Hui mapped things that way. I will double check to see the rationals behind that.

**Comment by b-sumner** [comment]
The clock functions are OK, although I would prefer an explicit cast when discarding the upper half of the memrealtime.

**Comment by guansong** [comment]
I spent some time try to further split the wrapper functions from the original file, I can see other things like

1. workitem functions
2. bits counting functions
3. minum on int 

are in the same situations as barrier. I will ask Greg about this in our meeting. 

Looks like PR has to be closed. But, I will put a few new functions here to see how do you prefer them to be added.

File: cuda2gcn/src/lanemask.ll [hunk 1]
```diff
@@ -0,0 +1,45 @@
+target datalayout = "e-p:64:64-p1:64:64-p2:64:64-p3:32:32-p4:64:64-p5:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64"
+target triple = "amdgcn--amdhsa"
+
+; Function Attrs: alwaysinline nounwind
+declare i32 @__ockl_activelane_u32() #0
+
+; Function Attrs: alwaysinline nounwind
+define i64 @__gcn_ballot64(i32 %a) #0 {
```
Comment (b-sumner) [inline cuda2gcn/src/lanemask.ll]: Implement this in C using __llvm_amdgcn_icmp_i32 from irif.h.  The last argument indicates which comparison.  ICMP_EQ=32, ICMP_NE=33, ICMP_UGT=34...  See InstrTypes.h in include/llvm/IR

File: cuda2gcn/src/lanemask.ll [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  %b = tail call i64 asm "v_cmp_ne_i32_e64 $0, 0, $1", "=s,v"(i32 %a) #1
+  ret i64 %b
+}
+
+; Function Attrs: alwaysinline nounwind
+define i64 @__gcn_lanemask_gt() local_unnamed_addr #0 {
```
Comment (b-sumner) [inline cuda2gcn/src/lanemask.ll]: Implement this in C.

File: cuda2gcn/src/lanemask.ll [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  %and = and i64 %shl, %call1
+  br label %cond.end
+
+cond.end:                                         ; preds = %entry, %cond.false
+  %cond = phi i64 [ %and, %cond.false ], [ 0, %entry ]
+  ret i64 %cond
+}
+
+; Function Attrs: alwaysinline nounwind
+define i64 @__gcn_lanemask_lt() local_unnamed_addr #0 {
```
Comment (b-sumner) [inline cuda2gcn/src/lanemask.ll]: Implement this in C

**Comment by guansong** [comment]
Thanks. I will check that direction. I will definitely close this PR, but before that, let's get the best use of it, I like to see your opinion on one more function.

File: cuda2gcn/src/trap.ll [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+; ModuleID = 'trap.bc'
+
+source_filename = "trap.ll"
+
+target datalayout = "e-p:64:64-p1:64:64-p2:64:64-p3:32:32-p4:64:64-p5:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64"
+target triple = "amdgcn--amdhsa"
+
+; Function Attrs: alwaysinline noreturn nounwind
+define void @__device_trap() #0 {
```
Comment (b-sumner) [inline cuda2gcn/src/trap.ll]: This can be implemented in C.  Just call __builtin_abort().

**Comment by guansong** [comment]
I see. Thanks!

**Pull request closed without merging**